### DISCUSSION TITLE
Missing commands for NICE Spa - IBT4ZWAVE - BusT4-Z-Wave interface

### DISCUSSION THREAD

**Discussion by stomko11** [zwave-js/zwave-js#4804] [General] [answered]
Support for NICE Spa - IBT4ZWAVE - BusT4-Z-Wave interface has been added several months back. However I struggle to be able to control it properly. In Fibaro HC3, I was able to open/close the gate and stop it at any position.
Here, it looks like that only available parameters in multilevel switch are Current and Target value, which allow only values 0 or 99 to close/open the gate. I was not able to find any way to stop the gate in any position, or specify value (similar to when using cover).

Here is the ZwaveJS DB device: https://devices.zwave-js.io/?jumpTo=0x0441:0x2400:0x1000:0.0
PR where device was added: https://github.com/zwave-js/node-zwave-js/pull/2525
And here are the commands supported by the device: https://www.niceforyou.com/sites/default/files/upload/manuals/isa00mm.pdf

If I look at "**Table 6 - Multilevel Switch Command Class SET**" it does not look like there is any command that can be sent to stop the gate.

Any advise would be appreciated.

**Comment by kpine** [comment, marked as answer]
You could contact the manufacturer and ask how to do it.

If it's anything like normal window covers, you can use the Multilevel Switch Stop Level Change Command. All devices should support this command. There's a CC API method to do this, [`stopLevelChange`](https://zwave-js.github.io/node-zwave-js/#/api/CCs/MultilevelSwitch?id=stoplevelchange). 

If you're using the Value API, such as in zwavejs2mqtt or HA, either of these `Up` or `Down` values will send the stop command when you use an "Off" (`false`) value. If you don't have Up/Down, it could be [something else](https://github.com/zwave-js/node-zwave-js/blob/f288e7fbf2e168cc5ce1d862f090867a6416e589/packages/zwave-js/src/lib/commandclass/_Types.ts#L1083-L1092) like `Close/Open`.

![image](https://user-images.githubusercontent.com/791794/179023611-e7f7063b-0659-492e-8d86-813de44ac37b.png)

If not `stopLevelChange`, then sometimes you have to set the multilevel targetValue based on the direction. If it's opening, send close (0), and if it's closing, send open (99).

**Comment by Loic691** [comment]
Hi
Yes we can send with zwave-js-ui the Stop command in the 38/0/targetValue/set command to stop it
So ther is a problem in the 38/0/currentValue command wich now get only 0 or 99 (seams closed or opened).
In fact manufactuer says the 254 value is a opening, closing or Stopped status..
Here's the manufacturer  data
![image](https://user-images.githubusercontent.com/60837526/196919296-3655dbde-60f9-4a08-b01b-c2130c89e932.png)

The duration state is an error too.
When closed we have duration = 0
When opened we have "unknown" instead of 254 (0xFE)

I think there is a bug wich value > 99 in the zwave-js driver

**Reply by AlCalzone** [reply]
> When opened we have "unknown" instead of 254 (0xFE)

Yes because this is the definition of that value from the Z-Wave specification. Your device doesn't  seem to have defined intermediate positions, so the current value is unknown in these cases.

**Reply by Loic691** [reply]
Ok so i will try to use the unknow value of duration to compute the good state.
So I understand that Nice have worse implementation of the zwave protocol. I am right ?

**Reply by AlCalzone** [reply]
I wouldn't say it's bad, they are using a feature that the protocol provides. 

This seems more like a hardware or implementation limitation, other shutters seem to interpolate the current position by moving time instead of just saying the position is unknown.

**Reply by Loic691** [reply]
ok understand

**Comment by arnog23** [comment]
Hi,

From what I understand from the [Zwave specification](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwiChsHmvND7AhWnQaQEHQQlBt0QFnoECAoQAQ&url=https%3A%2F%2Fsdomembers.z-wavealliance.org%2Fdocument%2Fdl%2F634&usg=AOvVaw027MEDnDihIxMtsd3a80A9) (page 100), value 254 is allowed as "Unknown position" for the Multilevel Switch CC.

![image](https://user-images.githubusercontent.com/60715520/204229921-449807b6-9902-45fc-8e9a-8359eaa5dd4d.png)

whereas ZwaveJS is not allowing it as the value is restricted between 0 and 99.

Would it be possible to allow 254 value for the currentValue and targetValue in the Multilevel Switch CC ?

This 254 value was allowed by OpenZwave and it was helpfull to know if the gate is opening, stopped or closing and not only opened (99) or closed (0).

Thanks.

**Reply by Loic691** [reply]
Yes i'm agree with that.
It's a currently issue...

**Reply by AlCalzone** [reply]
"unknown" values are translated to `undefined` (default) or `"unknown"`, depending on the driver option `preserveUnknownValues`.

Not sure how transitions to `undefined` get handled by zwave-js-server and zwave-js-ui, but the driver does support "unknown" values.

@raman325 @robertsLando ?

**Reply by robertsLando** [reply]
Regarding value updates It is treated like any other value update so the valueId value is set to `undefined` or `"unknown"` based on what the driver sends me.

About setValue api I'm not sure UI allows to send 'undefined' values, it may send `''` instead. Also setValue api try to coerce the value sent to Number when the valueId is type `number` https://github.com/zwave-js/zwave-js-ui/blob/master/lib/ZwaveClient.ts#L2663

**Reply by raman325** [reply]
`undefined` doesn't serialize so it's effectively a `null` for the purposes of the HA integration. `"unknown"` does get serialized so we receive it in the python library, but we convert it to a `null`. So the end result is that there is no distinction between the two. When HA gets a `null`, the state is `"unknown"` so both `undefined` and `"unknown"` will result in state `"unknown"` in HA.

**Reply by AlCalzone** [reply]
Ok so in other words: users already get the unknown state in HA? Or would the driver option have to be enabled first?

**Reply by raman325** [reply]
`undefined` and `"unknown"` both result in an `"unknown"` state in HA. I believe that means it is irrespective of the driver option

**Reply by AlCalzone** [reply]
Cool. Then I don't know what the issue here is tbh 😅

**Reply by arnog23** [reply]
So if I understand well, from ZWaveJS point of view, the value is based on what the driver sends to it but I am not sure to understand what is done at ZwaveJS UI level (I am not using HA).

**Reply by robertsLando** [reply]
> the value is based on what the driver sends to it 

same in zwavejs ui

**Reply by AlCalzone** [reply]
In zwave-js-ui, there should be **no value** (if I got this right) when the value is 254/unknown.

**Reply by arnog23** [reply]
> In zwave-js-ui, there should be **no value** (if I got this right) when the value is 254/unknown.

So what should be done to get this 254 value instead of the "no value" ? Set the driver option preserveUnknownValues (at zwave-js level) to True ?

https://zwave-js.github.io/node-zwave-js/#/api/driver?id=driver-methods

![image](https://user-images.githubusercontent.com/60715520/204542691-01d78be7-19c6-4a1e-b2ad-786805fd774d.png)

I understand that when this option is configured to True, the value will be exposed  as the literal "unknown" even if the value is normally numeric. Is there a way to change this behaivour in order to get the real numeric value ?

**Reply by robertsLando** [reply]
This is just a flag to send `unknown` as string instead of `undefined` so it's not what you are looking for.

**Reply by AlCalzone** [reply]
You won't get the raw value 254. Not sure what exactly you want to do with it, since **no value** or **unknown** indicate the same state.

**Reply by arnog23** [reply]
So what is the way to get this numeric value ?

EDIT : I missed AlCalzone answer before writing my answer.

**Reply by arnog23** [reply]
> You won't get the raw value 254. Not sure what exactly you want to do with it, since **no value** or **unknown** indicate the same state.

I would like to use this value to be able to get the 5 different states (open, opening, stopped, closing, closed) as indicated in the Bidi documentation and not only the 2 default (opened or closed).

Actually, when looking at zwave-js-ui level, currentValue and targetValue never get the value 254 or unknown. They only get 0 or 99.

If we were at least able to get the unknown value, we will be fine with it (knowing that it means 254) but we do not even get it.

**Reply by AlCalzone** [reply]
> Actually, when looking at zwave-js-ui level, currentValue and targetValue never get the value 254 or unknown. They only get 0 or 99.

Can you show both application logs and driver logs (level debug) for a situation where the value should become 254/unknown?

**Reply by arnog23** [reply]
Here is what I got when setting both application and driver logs to debug level and filtered on node 2.

The gate was closed. I have pushed the open button so from the Bidi specification, we should have currentValue changing from 0 to 254 and targetValue changing from 0 to 99.

then, I pushed the stop button, so currentValue should stay to 254 and targetValue should change from 99 to 254.

![image](https://user-images.githubusercontent.com/60715520/204896194-3e104e51-fa7d-433c-8c07-a09ee6feab72.png)


But from the logs, we are only seeing the targetValue changing from 0 to 99 when I have pushed the open button.

Could the value change to 254 this line in the driver logs ? `"current value":"undefined`

```
"2022-11-30 20":"23":"56.747 INFO Z-WAVE":"Node 2":"value updated":"38-0-targetValue 0 => 99
2022-11-30 20":"23":"56.747 DEBUG MQTT":"Publishing to zwave/_EVENTS/ZWAVE_GATEWAY-Jeedom/node/node_value_updated":{
   "data":[
      {
         "id":2,
         "inited":true,
         "name":"undefined",
         "location":"undefined",
         "status":4,
         "isControllerNode":false,
         "interviewStage":5,
         "deviceClass":"DeviceClass"{
            "basic":{
               "key":4,
               "label":"Routing Slave"
            },
            "generic":"GenericDeviceClass"{
               "key":17,
               "label":"Multilevel Switch",
               "supportedCCs":[
                  "Array"
               ],
               "controlledCCs":[
                  "Array"
               ],
               "specific":[
                  "Map"
               ]
            },
            "specific":"SpecificDeviceClass"{
               "key":0,
               "label":"Unused",
               "supportedCCs":[
                  "Array"
               ],
               "controlledCCs":[
                  "Array"
               ]
            },
            "_mandatorySupportedCCs":[
               32,
               38,
               [
                  "length"
               ]:2
            ],
            "_mandatoryControlledCCs":[
               [
                  "length"
               ]:0
            ],
            [
               "mandatorySupportedCCs"
            ]:[
               "Getter"
            ],
            [
               "mandatoryControlledCCs"
            ]:[
               "Getter"
            ]
         },
         "zwavePlusVersion":2,
         "ready":true,
         "zwavePlusRoleType":5,
         "isListening":true,
         "isFrequentListening":false,
         "canSleep":false,
         "isRouting":true,
         "supportedDataRates":[
            40000,
            100000,
            [
               "length"
            ]:2
         ],
         "maxDataRate":100000,
         "supportsSecurity":false,
         "isSecure":false,
         "supportsBeaming":true,
         "protocolVersion":3,
         "sdkVersion":"7.13.1",
         "firmwareVersion":"7.0.0",
         "manufacturerId":1089,
         "manufacturer":"NICE Spa",
         "productId":4096,
         "productDescription":"BusT4-Z-Wave interface",
         "productType":9216,
         "productLabel":"IBT4ZWAVE",
         "deviceDatabaseUrl":"https://devices.zwave-js.io/?jumpTo=0x0441:0x2400:0x1000:7.0.0",
         "keepAwake":false
      },
      {
         "commandClassName":"Multilevel Switch",
         "commandClass":38,
         "property":"targetValue",
         "propertyKey":"undefined",
         "endpoint":0,
         "newValue":99,
         "prevValue":0,
         "propertyName":"targetValue"
      },
      [
         "length"
      ]:2
   ]
}"with options"{
   "qos":1,
   "retain":false
}"2022-11-30T19":"23":56.747Z CNTRLR[
   Node 002
][
   "~"
][
   "Multilevel Switch"
]"duration":"unknown""=>""unknow [Endpoint 0]
n""2022-11-30 20":"23":"56.747 DEBUG MQTT":"Publishing to zwave/2/38/0/duration":{
   "time":1669836236747,
   "value":"Duration"{
      "unit":"unknown",
      "_value":0,
      [
         "value"
      ]:[
         "Getter/Setter"
      ]
   }
}"with options"{
   "qos":1,
   "retain":true
}"2022-11-30 20":"23":"56.747 INFO Z-WAVE":"Node 2":"value updated":"38-0-duration unknown => unknown
2022-11-30 20":"23":"56.748 DEBUG MQTT":"Publishing to zwave/_EVENTS/ZWAVE_GATEWAY-Jeedom/node/node_value_updated":{
   "data":[
      {
         "id":2,
         "inited":true,
         "name":"undefined",
         "location":"undefined",
         "status":4,
         "isControllerNode":false,
         "interviewStage":5,
         "deviceClass":"DeviceClass"{
            "basic":{
               "key":4,
               "label":"Routing Slave"
            },
            "generic":"GenericDeviceClass"{
               "key":17,
               "label":"Multilevel Switch",
               "supportedCCs":[
                  "Array"
               ],
               "controlledCCs":[
                  "Array"
               ],
               "specific":[
                  "Map"
               ]
            },
            "specific":"SpecificDeviceClass"{
               "key":0,
               "label":"Unused",
               "supportedCCs":[
                  "Array"
               ],
               "controlledCCs":[
                  "Array"
               ]
            },
            "_mandatorySupportedCCs":[
               32,
               38,
               [
                  "length"
               ]:2
            ],
            "_mandatoryControlledCCs":[
               [
                  "length"
               ]:0
            ],
            [
               "mandatorySupportedCCs"
            ]:[
               "Getter"
            ],
            [
               "mandatoryControlledCCs"
            ]:[
               "Getter"
            ]
         },
         "zwavePlusVersion":2,
         "ready":true,
         "zwavePlusRoleType":5,
         "isListening":true,
         "isFrequentListening":false,
         "canSleep":false,
         "isRouting":true,
         "supportedDataRates":[
            40000,
            100000,
            [
               "length"
            ]:2
         ],
         "maxDataRate":100000,
         "supportsSecurity":false,
         "isSecure":false,
         "supportsBeaming":true,
         "protocolVersion":3,
         "sdkVersion":"7.13.1",
         "firmwareVersion":"7.0.0",
         "manufacturerId":1089,
         "manufacturer":"NICE Spa",
         "productId":4096,
         "productDescription":"BusT4-Z-Wave interface",
         "productType":9216,
         "productLabel":"IBT4ZWAVE",
         "deviceDatabaseUrl":"https://devices.zwave-js.io/?jumpTo=0x0441:0x2400:0x1000:7.0.0",
         "keepAwake":false
      },
      {
         "commandClassName":"Multilevel Switch",
         "commandClass":38,
         "property":"duration",
         "propertyKey":"undefined",
         "endpoint":0,
         "newValue":"Duration"{
            "unit":"unknown",
            "_value":0,
            [
               "value"
            ]:[
               "Getter/Setter"
            ]
         },
         "prevValue":"Duration"{
            "unit":"unknown",
            "_value":0,
            [
               "value"
            ]:[
               "Getter/Setter"
            ]
         },
         "propertyName":"duration"
      },
      [
         "length"
      ]:2
   ]
}"with options"{
   "qos":1,
   "retain":false
}"2022-11-30T19":"23":56.749Z DRIVER «[
   Node 002
][
   "REQ"
][
   "ApplicationCommand"
]"└─"[
   "MultilevelSwitchCCReport"
]"current value":"undefined
target value":"99
duration":"unknown
"2022-11-30 20":"24":"00.791 INFO Z-WAVE":"Node 2":"value updated":"38-0-duration unknown => 0s
2022-11-30 20":"24":"00.791 DEBUG MQTT":"Publishing to zwave/_EVENTS/ZWAVE_GATEWAY-Jeedom/node/node_value_updated":{
   "data":[
      {
         "id":2,
         "inited":true,
         "name":"undefined",
         "location":"undefined",
         "status":4,
         "isControllerNode":false,
         "interviewStage":5,
         "deviceClass":"DeviceClass"{
            "basic":{
               "key":4,
               "label":"Routing Slave"
            },
            "generic":"GenericDeviceClass"{
               "key":17,
               "label":"Multilevel Switch",
               "supportedCCs":[
                  "Array"
               ],
               "controlledCCs":[
                  "Array"
               ],
               "specific":[
                  "Map"
               ]
            },
            "specific":"SpecificDeviceClass"{
               "key":0,
               "label":"Unused",
               "supportedCCs":[
                  "Array"
               ],
               "controlledCCs":[
                  "Array"
               ]
            },
            "_mandatorySupportedCCs":[
               32,
               38,
               [
                  "length"
               ]:2
            ],
            "_mandatoryControlledCCs":[
               [
                  "length"
               ]:0
            ],
            [
               "mandatorySupportedCCs"
            ]:[
               "Getter"
            ],
            [
               "mandatoryControlledCCs"
            ]:[
               "Getter"
            ]
         },
         "zwavePlusVersion":2,
         "ready":true,
         "zwavePlusRoleType":5,
         "isListening":true,
         "isFrequentListening":false,
         "canSleep":false,
         "isRouting":true,
         "supportedDataRates":[
            40000,
            100000,
            [
               "length"
            ]:2
         ],
         "maxDataRate":100000,
         "supportsSecurity":false,
         "isSecure":false,
         "supportsBeaming":true,
         "protocolVersion":3,
         "sdkVersion":"7.13.1",
         "firmwareVersion":"7.0.0",
         "manufacturerId":1089,
         "manufacturer":"NICE Spa",
         "productId":4096,
         "productDescription":"BusT4-Z-Wave interface",
         "productType":9216,
         "productLabel":"IBT4ZWAVE",
         "deviceDatabaseUrl":"https://devices.zwave-js.io/?jumpTo=0x0441:0x2400:0x1000:7.0.0",
         "keepAwake":false
      },
      {
         "commandClassName":"Multilevel Switch",
         "commandClass":38,
         "property":"duration",
         "propertyKey":"undefined",
         "endpoint":0,
         "newValue":"Duration"{
            "unit":"seconds",
            "_value":0,
            [
               "value"
            ]:[
               "Getter/Setter"
            ]
         },
         "prevValue":"Duration"{
            "unit":"unknown",
            "_value":0,
            [
               "value"
            ]:[
               "Getter/Setter"
            ]
         },
         "propertyName":"duration"
      },
      [
         "length"
      ]:2
   ]
}"with options"{
   "qos":1,
   "retain":false
}"2022-11-30T19":"24":00.792Z DRIVER «[
   Node 002
][
   "REQ"
][
   "ApplicationCommand"
]"└─"[
   "MultilevelSwitchCCReport"
]"current value":"undefined
```

**Reply by robertsLando** [reply]
@arnog23 that are just zui logs, he needs [driver logs](https://zwave-js.github.io/zwave-js-ui/#/troubleshooting/generating-logs?id=generating-log-files)

**Reply by arnog23** [reply]
I have configured Application and Driver to Debug level. Do I need to configure them both with Silly level ?

**Reply by robertsLando** [reply]
@arnog23 if you are logging to file the driver doesn't also log to stdout, you should attach here the file with prefix `zwavejs-XXX.log`

**Reply by arnog23** [reply]
```
2022-12-01T09:46:14.616Z CNTRLR   [Node 002] Embedded device config loaded
2022-12-01T09:46:15.410Z CNTRLR » [Node 002] pinging the node...
2022-12-01T09:46:16.127Z DRIVER » [Node 002] [REQ] [SendData]
                                  │ transmit options: 0x25
                                  │ callback id:      1
                                  └─[NoOperationCC]
2022-12-01T09:46:16.318Z SERIAL « 0x01180004000a12600d0103320221340000006e000900000072a4              (26 bytes)
2022-12-01T09:46:16.319Z SERIAL » [ACK]                                                                   (0x06)
2022-12-01T09:46:16.319Z SERIAL « [CAN]                                                                   (0x18)
2022-12-01T09:46:16.421Z DRIVER » [Node 002] [REQ] [SendData]
                                  │ transmit options: 0x25
                                  │ callback id:      1
                                  └─[NoOperationCC]
2022-12-01T09:46:16.423Z SERIAL « [ACK]                                                                   (0x06)
2022-12-01T09:46:16.434Z SERIAL « 0x0104011301e8                                                       (6 bytes)
2022-12-01T09:46:16.434Z SERIAL » [ACK]                                                                   (0x06)
2022-12-01T09:46:16.435Z DRIVER « [RES] [SendData]
                                    was sent: true
2022-12-01T09:46:16.450Z SERIAL « 0x0107001301000002e8                                                 (9 bytes)
2022-12-01T09:46:16.450Z SERIAL » [ACK]                                                                   (0x06)
2022-12-01T09:46:16.451Z DRIVER « [REQ] [SendData]
                                    callback id:     1
                                    transmit status: OK
2022-12-01T09:46:16.456Z CNTRLR   [Node 002] The node is alive.
2022-12-01T09:46:16.482Z CNTRLR   [Node 002] The node is ready to be used
2022-12-01T09:46:16.484Z CNTRLR « [Node 002] ping successful
2022-12-01T09:46:16.485Z SERIAL » 0x010800130401002502c6                                              (10 bytes)
2022-12-01T09:46:47.008Z CNTRLR   [Node 002] [translateValueEvent: value updated]
                                    commandClass: Multilevel Switch
                                    endpoint:     0
                                    property:     targetValue
                                    propertyKey:  undefined
                                    internal:     false
                                    secret:       false
                                    event source: undefined
2022-12-01T09:46:47.008Z CNTRLR   [Node 002] [~] [Multilevel Switch] targetValue: 0 => 99           [Endpoint 0]
2022-12-01T09:46:47.008Z CNTRLR   [Node 002] [translateValueEvent: value updated]
                                    is root endpoint:        true
                                    is application CC:       true
                                    should hide root values: false
2022-12-01T09:46:47.009Z CNTRLR   [Node 002] [translateValueEvent: value updated]
                                    commandClass: Multilevel Switch
                                    endpoint:     0
                                    property:     duration
                                    propertyKey:  undefined
                                    internal:     false
                                    secret:       false
                                    event source: undefined
2022-12-01T09:46:47.009Z CNTRLR   [Node 002] [~] [Multilevel Switch] duration: {"value":0,"unit":"s [Endpoint 0]
                                  econds"} => "unknown"
2022-12-01T09:46:47.009Z CNTRLR   [Node 002] [translateValueEvent: value updated]
                                    is root endpoint:        true
                                    is application CC:       true
                                    should hide root values: false
2022-12-01T09:46:47.011Z DRIVER « [Node 002] [REQ] [ApplicationCommand]
                                  └─[MultilevelSwitchCCReport]
                                      current value: undefined
                                      target value:  99
                                      duration:      unknown
2022-12-01T09:46:47.824Z SERIAL « 0x0118000400bc12600d0103320221340000003a00090000003703              (26 bytes)
2022-12-01T09:46:47.826Z SERIAL » [ACK]                                                                   (0x06)
2022-12-01T09:46:55.662Z CNTRLR   [Node 002] [translateValueEvent: value updated]
                                    commandClass: Multilevel Switch
                                    endpoint:     0
                                    property:     duration
                                    propertyKey:  undefined
                                    internal:     false
                                    secret:       false
                                    event source: undefined
2022-12-01T09:46:55.663Z CNTRLR   [Node 002] [~] [Multilevel Switch] duration: "unknown" => {"value [Endpoint 0]
                                  ":0,"unit":"seconds"}
2022-12-01T09:46:55.663Z CNTRLR   [Node 002] [translateValueEvent: value updated]
                                    is root endpoint:        true
                                    is application CC:       true
                                    should hide root values: false
2022-12-01T09:46:55.664Z DRIVER « [Node 002] [REQ] [ApplicationCommand]
                                  └─[MultilevelSwitchCCReport]
                                      current value: undefined
2022-12-01T09:46:56.764Z SERIAL « 0x0118000400bc12600d0103320221340000003600090000003a02              (26 bytes)
2022-12-01T09:46:56.765Z SERIAL » [ACK]                                                                   (0x06)

```

Could it be here where current value is `undefined` and not getting 254 value ?

```
2022-12-01T09:46:47.011Z DRIVER « [Node 002] [REQ] [ApplicationCommand]
                                  └─[MultilevelSwitchCCReport]
                                      current value: undefined
                                      target value:  99
                                      duration:      unknown
```

**Reply by arnog23** [reply]
Are those logs helpful ?

**Reply by AlCalzone** [reply]
It looks like the `undefined` value change doesn't get emitted. This is something I will have to investigate.

@robertsLando can the UI handle if a number gets set to "unknown"? In that case, they could enable the driver option as a workaround (can you explain how?).

**Reply by robertsLando** [reply]
I don't have any check in code that skips undefined values, if I receive a value update with a value that is undefined I treat it like others. About enabling the driver option they have to manually edit settings.json file and add that option under `zwave.options`. This can be easily done from store explorer

**Reply by arnog23** [reply]
Here is what I have when adding `"options":{"preserveUnknownValues":true}` in settings.json.

Closed:

![PreserveUnknownClosed](https://user-images.githubusercontent.com/60715520/205434263-16868899-22e4-4f97-85d7-4129185f22f1.png)

Opening:
![PreserveUnknownOpening](https://user-images.githubusercontent.com/60715520/205434273-5d054f6a-3ba9-4fb6-9f8e-f14fce2c8ed0.png)

Opening Stoppped:
![PreserveUnnownOpeningStopped](https://user-images.githubusercontent.com/60715520/205434282-b19ac587-b113-4b34-aa29-1f537a7bc8cf.png)

Open:
![PreserveUnknownOpened](https://user-images.githubusercontent.com/60715520/205434293-99646ee8-5df0-4f29-a2f7-d10e30915b88.png)

ClosingStopped:
![PreserveUnknownClosingStopped](https://user-images.githubusercontent.com/60715520/205434303-a6cdb55d-c5ca-4f8d-8deb-99ee24b15c24.png)

Closed:
![PreserveUnknownClosed](https://user-images.githubusercontent.com/60715520/205434342-70db349a-167d-4f58-8454-5a305a50d40a.png)

Which is close to the Bidi specification except for the stopped status where targetValue is not set to 254 but it is may be better this way as it allows to know if it is a stop when closing or a stop when opening.

![image](https://user-images.githubusercontent.com/60715520/205434799-6a9396a7-6335-4076-8634-cbd94dab42b3.png)


```
2022-12-03T09:41:18.994Z CNTRLR   [Node 002] Embedded device config loaded
2022-12-03T09:41:19.630Z CNTRLR » [Node 002] pinging the node...
2022-12-03T09:41:19.971Z DRIVER » [Node 002] [REQ] [SendData]
                                  │ transmit options: 0x25
                                  │ callback id:      1
                                  └─[NoOperationCC]
2022-12-03T09:41:20.076Z SERIAL « [ACK]                                                                   (0x06)
2022-12-03T09:41:20.077Z SERIAL « 0x0104011301e8                                                       (6 bytes)
2022-12-03T09:41:20.078Z SERIAL » [ACK]                                                                   (0x06)
2022-12-03T09:41:20.089Z DRIVER « [RES] [SendData]
                                    was sent: true
2022-12-03T09:41:20.107Z SERIAL « 0x0107001301000007ed                                                 (9 bytes)
2022-12-03T09:41:20.108Z SERIAL » [ACK]                                                                   (0x06)
2022-12-03T09:41:20.120Z DRIVER « [REQ] [SendData]
                                    callback id:     1
                                    transmit status: OK
2022-12-03T09:41:20.126Z CNTRLR   [Node 002] The node is alive.
2022-12-03T09:41:20.138Z CNTRLR   [Node 002] The node is ready to be used
2022-12-03T09:41:20.138Z CNTRLR « [Node 002] ping successful
2022-12-03T09:41:20.139Z SERIAL » 0x010800130401002502c6                                              (10 bytes)
2022-12-03T09:41:55.468Z CNTRLR   [Node 002] [translateValueEvent: value updated]
                                    commandClass: Multilevel Switch
                                    endpoint:     0
                                    property:     duration
                                    propertyKey:  undefined
                                    internal:     false
                                    secret:       false
                                    event source: undefined
2022-12-03T09:41:55.468Z CNTRLR   [Node 002] [~] [Multilevel Switch] duration: {"value":0,"unit":"s [Endpoint 0]
                                  econds"} => {"value":0,"unit":"seconds"}
2022-12-03T09:41:55.468Z CNTRLR   [Node 002] [translateValueEvent: value updated]
                                    is root endpoint:        true
                                    is application CC:       true
                                    should hide root values: false
2022-12-03T09:41:55.469Z CNTRLR   [Node 002] [translateValueEvent: value updated]
                                    commandClass: Multilevel Switch
                                    endpoint:     0
                                    property:     currentValue
                                    propertyKey:  undefined
                                    internal:     false
                                    secret:       false
                                    event source: undefined
2022-12-03T09:41:55.469Z CNTRLR   [Node 002] [~] [Multilevel Switch] currentValue: 0 => "unknown"   [Endpoint 0]
2022-12-03T09:41:55.469Z CNTRLR   [Node 002] [translateValueEvent: value updated]
                                    is root endpoint:        true
                                    is application CC:       true
                                    should hide root values: false
2022-12-03T09:41:55.470Z DRIVER « [Node 002] [REQ] [ApplicationCommand]
                                  └─[MultilevelSwitchCCReport]
                                      current value: unknown
2022-12-03T09:41:56.032Z SERIAL « 0x010c000400dd0631050422000f31                                      (14 bytes)
2022-12-03T09:41:56.033Z SERIAL » [ACK]                                                                   (0x06)
2022-12-03T09:41:59.384Z CNTRLR   [Node 002] [translateValueEvent: value updated]
                                    commandClass: Multilevel Switch
                                    endpoint:     0
                                    property:     targetValue
                                    propertyKey:  undefined
                                    internal:     false
                                    secret:       false
                                    event source: undefined
2022-12-03T09:41:59.384Z CNTRLR   [Node 002] [~] [Multilevel Switch] targetValue: 0 => 0            [Endpoint 0]
2022-12-03T09:41:59.384Z CNTRLR   [Node 002] [translateValueEvent: value updated]
                                    is root endpoint:        true
                                    is application CC:       true
                                    should hide root values: false
2022-12-03T09:41:59.384Z CNTRLR   [Node 002] [translateValueEvent: value updated]
                                    commandClass: Multilevel Switch
                                    endpoint:     0
                                    property:     duration
                                    propertyKey:  undefined
                                    internal:     false
                                    secret:       false
                                    event source: undefined
2022-12-03T09:41:59.384Z CNTRLR   [Node 002] [~] [Multilevel Switch] duration: {"value":0,"unit":"s [Endpoint 0]
                                  econds"} => "unknown"
2022-12-03T09:41:59.384Z CNTRLR   [Node 002] [translateValueEvent: value updated]
                                    is root endpoint:        true
                                    is application CC:       true
                                    should hide root values: false
2022-12-03T09:41:59.385Z CNTRLR   [Node 002] [translateValueEvent: value updated]
                                    commandClass: Multilevel Switch
                                    endpoint:     0
                                    property:     currentValue
                                    propertyKey:  undefined
                                    internal:     false
                                    secret:       false
                                    event source: undefined
2022-12-03T09:41:59.385Z CNTRLR   [Node 002] [~] [Multilevel Switch] currentValue: "unknown" => "un [Endpoint 0]
                                  known"
2022-12-03T09:41:59.385Z CNTRLR   [Node 002] [translateValueEvent: value updated]
                                    is root endpoint:        true
                                    is application CC:       true
                                    should hide root values: false
2022-12-03T09:41:59.385Z DRIVER « [Node 002] [REQ] [ApplicationCommand]
                                  └─[MultilevelSwitchCCReport]
                                      current value: unknown
                                      target value:  0
                                      duration:      unknown
2022-12-03T09:41:59.492Z SERIAL « 0x01180004002612600d0101320221340000002200090000002793              (26 bytes)
2022-12-03T09:41:59.493Z SERIAL » [ACK]                                                                   (0x06)
```
So it seems to looks good when adding this option `preserveUnknownValues` even if I would say it would be easier to have a numeric value (254) instead of a unknown value to have only numeric value instead of a mix of numeric and string values.

All above was done using physical buttons of the gate.

But when I use Zwave commands, I have the following issue:

open or close, then stop. When opening or closing again, I  have the following error message `Driver: options is not a MultilevelSwitchCCStartLevelChangeOptions (ZW0322)` and I am not able to control the Bidi anymore. I have to restart ZwaveJS.

This error do not appear when the option `preserveUnknownValues` is set to default (False).

**Reply by AlCalzone** [reply]
I think the error might be because the "unknown" duration is used for the command, which is nonsense.

 Could you put your last comment in an actual issue (not discussion)? Use the empty issue template (below the other choices).

**Reply by arnog23** [reply]
> Could you put your last comment in an actual issue (not discussion)? Use the empty issue template (below the other choices).

Here it is : [Driver: options is not a MultilevelSwitchCCStartLevelChangeOptions (ZW0322) / "options":{"preserveUnknownValues":true} #5237](https://github.com/zwave-js/node-zwave-js/issues/5237)

**Reply by arnog23** [reply]
Any chance to change the default behavior (without the `preserveUnknownValues` option) to have the numeric value 254 instead of the `unknown` or `undefined` string which would 

1. stick to the Zwave Alliance definition for the Multilevel Switch CC
2. avoid the `Driver: options is not a MultilevelSwitchCCStartLevelChangeOptions (ZW0322)`
3. let the user decide what to do with this 254 value. It is easier to compare only numeric values 0, 99 and 254 than a mix of numeric values 0 , 99 and a string unknown or undefined.

**Reply by AlCalzone** [reply]
I need to think about it.

On the other hand, "unknown" is much clearer in what it means than 254 to users who don't know the specs.

The issue under 2. won't be solved  by this change, since 254 is still not a valid value for set-type commands.

**Reply by arnog23** [reply]
> I need to think about it.

Thanks

> On the other hand, "unknown" is much clearer in what it means than 254 to users who don't know the specs.

Not really here. 254 means something else (opening, closing or stopped) than "unknown" based on combination of current and target values, so it have a signification. By the way, the other values are not translated as "open" (for 99) and "close" (for 0) but their numeric values are kept.

> The issue under 2. won't be solved by this change, since 254 is still not a valid value for set-type commands.

Ok

**Reply by arnog23** [reply]
Have you had time to think about it? :-)

**Reply by arnog23** [reply]
> > On the other hand, "unknown" is much clearer in what it means than 254 to users who don't know the specs.
> 
> Not really here. 254 means something else (opening, closing or stopped) than "unknown" based on combination of current and target values, so it have a signification. By the way, the other values are not translated as "open" (for 99) and "close" (for 0) but their numeric values are kept.

Any thought about it ?

**Reply by arnog23** [reply]
@AlCalzone, any chance to get improvement about this ?

**Reply by arnog23** [reply]
@AlCalzone, have you put it on your roadmap ?

**Comment by Stephane-80** [comment]
Yes 254 status value is very important because without it : 
0 -> closed 
99 -> open
without opening value (254) the system thinks that the door is fully closed even if it is 99% open in reality (sliding gate for example) but not at the end of its travel
the 0 value change to 99 only when the gate reaches full opening

**Comment by Seby9576** [comment]
Bonjour,

Il serait bien que le module IBT4 soit bien reconnu et les commandes créées automatiquement afin de recuperer l'etat en mouvement mais aussi l'etat des cellules par exemple

**Reply by AlCalzone** [reply]
English please 🙏🏻

**Comment by tetiti20** [comment]
I am also waiting for an update please

**Comment by jossalison38** [comment]
hello same problem for me, pedestrian opening sliding gate not taken into account.
I hope for progress on the subject
THANKS

**Comment by AlCalzone** [comment]
This is on my radar for v11. Locking this now since the "ping" spam is annoying.

**Reply by AlCalzone** [reply]
https://github.com/zwave-js/node-zwave-js/issues/5864
